<h1 style="text-align: center;">Introduction to Data Science - Assignment 10</h1>
<h3 style="text-align: center;">Name: Hugo Hayashi</h3>
<h3 style="text-align: center;">Student ID: 22446441</h3>



## Question 1


In [ ]:
# Your fancy code here
import numpy as np
import pandas as pd
from scipy.stats import norm
import statsmodels.api as sm
import requests, zipfile
import io

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00356/student.zip'

r = requests.get(url)

z = zipfile.ZipFile(io.BytesIO(r.content))
z.extractall()

df = pd.read_csv('student-mat.csv', sep=';')
df_dropped = df.dropna()
predictors = ['absences', 'health', 'freetime', 'studytime', 'traveltime']
for pred in predictors:
    model = sm.OLS(df_dropped['G1'], sm.add_constant(df_dropped[pred]))
    results = model.fit()
    print(f"\n\nLinear regression of {pred}:\n", results.summary())




Linear regression of absences:
                             OLS Regression Results                            
Dep. Variable:                     G1   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.3781
Date:                Tue, 24 Jun 2025   Prob (F-statistic):              0.539
Time:                        23:25:49   Log-Likelihood:                -1033.7
No. Observations:                 395   AIC:                             2071.
Df Residuals:                     393   BIC:                             2079.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.9

In [18]:
for pred in predictors: 
    model = sm.OLS(df_dropped['G1'], sm.add_constant(df_dropped[pred]))
    results = model.fit()
    print(f"\n\n{pred}:")
    print('\nCoefficients:', results.params[pred])
    print('\nStandard errors:', results.bse[pred])
    print('\nt-values:', results.tvalues[pred])
    print('\np-values:', results.pvalues[pred])
    print('\nR-squared:', results.rsquared)
    if results.pvalues[pred] < 0.05:
        print(f"\n\'{pred}\' is significant at the 5-percent level")
    else:
        print(f"\n\'{pred}\' is not significant at the 5-percent level")



absences:

Coefficients: -0.012858107369259802

Standard errors: 0.02091075058881693

t-values: -0.6149041525145595

p-values: 0.5389738820162011

R-squared: 0.0009611798672549687

'absences' is not significant at the 5-percent level


health:

Coefficients: -0.17469018429380714

Standard errors: 0.12010500462492076

t-values: -1.4544788107652296

p-values: 0.146611787321014

R-squared: 0.0053541523044793315

'health' is not significant at the 5-percent level


freetime:

Coefficients: 0.041912465545222155

Standard errors: 0.16760863642438753

t-values: 0.2500614910982223

p-values: 0.8026704586116559

R-squared: 0.00015908600906600068

'freetime' is not significant at the 5-percent level


studytime:

Coefficients: 0.6352199536555556

Standard errors: 0.19691328054241827

t-values: 3.2258868061401227

p-values: 0.001360987602330222

R-squared: 0.025796187386026292

'studytime' is significant at the 5-percent level


traveltime:

Coefficients: -0.4427465793227315

Standard errors: 0

Only studytime had a statistically significant and positive relationship with G1 while the other predictors showed little to no associations, indicating that study effort may be a more reliable predictor of academic performance than health, absences, freetime or traveltime.

## Question 2

In [20]:
#More code and comments...
X = df_dropped[predictors]

X = sm.add_constant(X)

model = sm.OLS(df_dropped.G1, X)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     G1   R-squared:                       0.037
Model:                            OLS   Adj. R-squared:                  0.025
Method:                 Least Squares   F-statistic:                     3.024
Date:                Tue, 24 Jun 2025   Prob (F-statistic):             0.0108
Time:                        23:43:30   Log-Likelihood:                -1026.3
No. Observations:                 395   AIC:                             2065.
Df Residuals:                     389   BIC:                             2089.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.4233      0.961     10.848      0.0

Constant: When all predictors are zero, the expected G1 score is 10.42. This effect is statistically significant.

Absences: Each additional absence reduces G1 by 0.009 points; however, this effect is marginal and not statistically significant.

Health: Each unit increase in health rating corresponds to a 0.15 point drop in G1. While counterintuitive, this effect is also not significant.

Freetime: Each unit increase in free time correspons to a 0.12 point increase in G1, but the result is not statistically significant.

Studytime: Each increase in studytime level increases G1 by 0.60 points on average. This is the only statistically significant predictor in the model.

Traveltime: Each unit increase in travel time corresponds to a 0.37 point decrease in G1. This effect is not statistically significant.

In [21]:
corr_matrix = df_dropped[predictors].corr()
corr_matrix

absences    health  freetime  studytime  traveltime
absences    1.000000 -0.029937 -0.058078  -0.062700   -0.012944
health     -0.029937  1.000000  0.075733  -0.075616    0.007501
freetime   -0.058078  0.075733  1.000000  -0.143198   -0.017025
studytime  -0.062700 -0.075616 -0.143198   1.000000   -0.100909
traveltime -0.012944  0.007501 -0.017025  -0.100909    1.000000

Not all of the coefficients are significant in the full model, and only studytime is statistically significant at the 5-percent level. The remaining predictors have p-values that exceed 0.05, making them not significant in this model.

In [22]:
X_reduced = sm.add_constant(df_dropped[['studytime']])
y = df_dropped['G1']
model_reduced = sm.OLS(y, X_reduced).fit()
print(model_reduced.summary())

                            OLS Regression Results                            
Dep. Variable:                     G1   R-squared:                       0.026
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     10.41
Date:                Tue, 24 Jun 2025   Prob (F-statistic):            0.00136
Time:                        23:43:34   Log-Likelihood:                -1028.7
No. Observations:                 395   AIC:                             2061.
Df Residuals:                     393   BIC:                             2069.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.6159      0.433     22.184      0.0

Although the full model has a slightly higher $R^2$, it includes four additional predictors that are not statistically significant. This complicates the model without adding real predictive value. Thus, despite its $R^2$ being lower, the reduced model is preferable as it is simpler, interpretable, avoids overfitting with variables that aren't informative, and balances complexity and performance, making it better for explaining the variance in G1 for practical use.

## Question 3

## Question 4